pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,738 @@
+				bridge:  ulStatus.Bridge,
+				netUUID: ulconfig.Network,
+			}
+			instData.bnNet[ulStatus.Bridge] = intfAttr
+
+			// build an App list cache, used for loop through all the Apps
+			if instData.appNet[status.AppNum] == nilUUID {
+				instData.appNet[status.AppNum] = status.UUIDandVersion.UUID
+				log.Infof("===FlowStats: appNet appNum %d, uuid %v\n", status.AppNum, instData.appNet[status.AppNum])
+			}
+
+			// build an acl cache indexed by app/aclnum, from flow MARK, we can get this aclAttr info
+			tmpMap := instData.ipaclattr[status.AppNum]
+			if tmpMap == nil {
+				tmpMap := make(map[int]aclAttr)
+				instData.ipaclattr[status.AppNum] = tmpMap
+			}
+			for _, rule := range ulStatus.ACLRules {
+				if rule.IsUserConfigured == false { // only include user defined rules
+					continue
+				}
+
+				var tempAttr aclAttr
+				tempAttr.aclNum = uint32(rule.RuleID)
+				tempAttr.chainName = rule.ActionChainName
+				tempAttr.aclName = rule.RuleName
+				tempAttr.tableName = rule.Table
+				tempAttr.vif = ulStatus.VifInfo.Vif
+				tempAttr.bridge = ulStatus.Bridge
+				if len(rule.Action) >= 2 { // '-j ACCEPT', '-j drop-all-bn1-nbu2x1'
+					tempAttr.action = rule.Action[1]
+				}
+				if _, ok := instData.ipaclattr[status.AppNum][int(rule.RuleID)]; !ok { // fake j as the aclNUM
+					instData.ipaclattr[status.AppNum][int(rule.RuleID)] = tempAttr
+				} else {
+					preAttr := instData.ipaclattr[status.AppNum][int(rule.RuleID)]
+					// the the entry exist, and already has the aclNum and bridge name, skip
+					if preAttr.aclNum == 0 || preAttr.bridge == "" {
+						instData.ipaclattr[status.AppNum][int(rule.RuleID)] = tempAttr
+					}
+				}
+			}
+		}
+	}
+}
+
+// DNSMonitor : DNS Query and Reply monitor on bridges
+func DNSMonitor(bn string, bnNum int) {
+	var (
+		err error
+		//action      string
+		snapshotLen int32 = 1280             // draft-madi-dnsop-udp4dns-00
+		promiscuous       = true             // mainly for switched network
+		timeout           = 10 * time.Second // collect enough packets in 10sec before processing
+		handle      *pcap.Handle
+		filter      = "udp and port 53"
+	)
+	if bnNum >= maxBridgeNumber {
+		log.Errorf("Can not snoop on brige number %d", bnNum)
+		return
+	}
+	log.Infof("(FlowStats) DNS Monitor on %s(bridge-num %d)", bn, bnNum)
+
+	handle, err = pcap.OpenLive(bn, snapshotLen, promiscuous, timeout)
+	if err != nil {
+		log.Errorf("Can not snoop on bridge %s", bn)
+		return
+	}
+	defer handle.Close()
+
+	err = handle.SetBPFFilter(filter)
+	if err != nil {
+		log.Errorf("Can not install DNS filter on %s", bn)
+		return
+	}
+
+	dnssys[bnNum].Done = make(chan bool)
+	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
+	dnsIn := packetSource.Packets()
+	for {
+		var packet gopacket.Packet
+		select {
+		case <-dnssys[bnNum].Done:
+			log.Infof("(FlowStats) DNS Monitor exit on %s(bridge-num %d)", bn, bnNum)
+			defer close(dnssys[bnNum].Done)
+			return
+		case packet = <-dnsIn:
```

Comment:
What happens if the interface is deleted by the kernel before you get the signal on the Don channel? Will it crash? log an error?

Reply:
I tested for deleting the bridge interface, and i expected to exit, but it does not. It does not crash certainly. That's why I put int the chan to get a signal